src/core/service.c

```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
                 log_unit_debug(UNIT(s), "Passing %zu fds to service", exec_params.n_socket_fds + exec_params.n_storage_fds);
         }
 
+        /* These options require PrivateUsers= when used in user units, as we need to be in a user namespace
+         * to have permission to enable them when not running as root. */
+        if (!s->exec_context.private_users && !MANAGER_IS_SYSTEM(UNIT(s)->manager))
+                s->exec_context.private_users =
+                                s->exec_context.private_tmp ||
+                                s->exec_context.private_devices ||
+                                s->exec_context.private_network ||
+                                s->exec_context.private_ipc ||
+                                s->exec_context.private_mounts ||
+                                s->exec_context.mount_apivfs ||
```

Comment:
all of the above should degrade "gracefully" if we can't apply them, hence we should *not* imply PrivateUsers= for them

Reply:
Which ones?

Reply:
Actually, I'm beginning to question this - why shouldn't all of these be made to work? What's the advantage of requiring extra hoops jumping for some options and not others? It would cause more confusion rather than less

Reply:
there are two kind of namespacing options:

1. sandboxing: these will hide things, make things unaccessible but not rearrange things. We generally make these options fail gracefully: if we cannot apply them due to lack of perms, lack of kernel support, running in a container or so, we'll simply not apply them. This should not break anything, it will just weaken protections.

2. options that rearrange stuff, such as BindPaths=, RootDirectory= and things. These options will rearrange the tree hiearchy, stuff that would normally appear at location X will appear at location Y, it's actually moved around, not just hidden or made unaccessible. If we can't apply these options, we'll fail starting the service.

I think for the second type of options it is fine to imply PrivateUsers=, if we don't the service will just fail. Hence there isn't really a compat break: making stuff work that previously didn't work is always OK. making stuff work differently that previously already worked in some other way would be a compat break.

For the first type of options otoh I think we should not imply PrivateUsers=. Because we already handle this gracefully: it's entirely OK to not apply them, and people could rely on that and have their stuff work. if we suddenly turn on privateUsers= for them, then things might end up working differently than before, and I'd call that a compat break.

Reply:
> For the first type of options otoh I think we should not imply PrivateUsers=. Because we already handle this gracefully: it's entirely OK to not apply them, and people could rely on that and have their stuff work. if we suddenly turn on privateUsers= for them, then things might end up working differently than before, and I'd call that a compat break.

But they always have to be explicitly requested? If you set "ReadOnlyPath=/foo", and /foo is read-only, then it's all working as intended? Also the line is not as clear - making a subtree inaccessible is not that different from bind mounting something over it, the filesystem view changes. Same for network namespaces, the network stack view changes, and so on. From the "compat" angle, this is a slight change in behaviour regardless, for all of these options. Might as well change them all, so that they make sense and just work as they should?

Reply:
PrivateUsers=yes means pretty much all users of the system (with the exception of root and the service's own user) will show up as "nobody", that's quite a change. We should not do that willy-nilly and pretend it was not a compat break...

hiding stuff and locking down access to stuff is systematically different from moving things to other places.

InaccessiblePath= and ReadOnlyPath= in a way only encode a contract: "hey, I won't need this stuff" or "hey, I am not gonna modify this". If systemd then follows up with making this read-only or making it disappear is not gonna change things as long as the contract is fulfilled by the service. Hence a graceful fallback makes sense, and has always been implemented.

But BindPaths= is systematically different: the contract sugegsts "hey, I expect to get resource X when I access path Y". we cannot just skip it ever, we have to fail if we can't make that happen, because the contract in this case implies "i am gonna use this" and not the opposite...

So to summarize:

1. Status quo: If ReadOnlyPath= cannot be executed via namespaces due to perms/containers/lack of kernel support we'll simply skip it, and nothing bad happens, "ls -al /home" will still look nice, with all home dirs owned by the users they belong to.

2. with you patch: If readonlypath= might not be executable, we'll allocate a userns with a mapping that hides all other users, and then make the specified dir read-only. But suddenly "ls -al /home" will look all bonkers, because everything is onwed by "nobody", and noone asked for it.

Reply:
A behaviour change is a behaviour change. The important point is, what's the cost vs benefit? On my Debian machine there are 0 references to ReadOnly/Protect/Inaccessible and so on under /usr/lib/systemd/user/. Do you see any on Fedora?
Doing the split you suggest is not cost-free either: it means keeping this weird split forever, where half the sandboxing options require an explicit unrelated configuration to be enabled, and half don't. Most of those always worked with PrivateUsers=yes, but nobody ever used them in user units, because nobody was aware they could work, because the setup is very confusing. Fixing this is a big win.

Reply:
By installing every package that provides a user unit I see the following:

- no Read/Write/Inaccessible path set anywhere
- no BindPath
- no RootDirectory/ExtensionDirectory
- one service using Protect* and Private* but it also already sets PrivateUsers (vvmd.service)
- one service using PrivateTmp without PrivateUsers (stunnel@.service)
- 4 services using ProtectSystem: mpd.service mpdscribble.service fluidsynth.service onedrive.service

So there's 5 units out of 284 that would be affected. I can probably get them fixed in a couple of days if needed, although looking at the comments I suspect they'd welcome it if this starts to work out of the box.

Reply:
I have created PRs for those 5 projects:

https://github.com/MusicPlayerDaemon/MPD/pull/1661
https://github.com/MusicPlayerDaemon/mpdscribble/pull/49
https://github.com/FluidSynth/fluidsynth/pull/1181
https://github.com/abraunegg/onedrive/pull/2205
https://salsa.debian.org/debian/stunnel/-/merge_requests/2 (downstream-only unit)

Reply:
All of the above have been fixed.
Found 2 more that are only shipped in Fedora:

https://gitlab.freedesktop.org/benzea/uresourced/-/merge_requests/24
https://github.com/abrt/reportd/pull/11

Reply:
Both of those were merged too

Reply:
Ack, they're stronger protections (against leg shooting).

Back to the `Private*` topic, I might have misunderstood inheritable&ambient capabilities (or they're initialized differently) because I tried the following (6.1.0-rc7) and it works:

```
$ unshare -U --keep-caps bash     # this is user@foo.service, PrivateUsers=yes
$ unshare -m bash                 # forked child (user service) creates mount ns
$ mount --bind /home /home -oro   # can set up protections as I wish, caps survive exec
```

I.e. it's not necessary to create userns in each user service but just run `user@.service` in a user namespace.

Reply:
I don't think we can enable a user namespace for the whole user session, too many things might break, as any and all programs ran by a user will be affected